_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+sets up the `LIBRARY_PATH` if necessary and finally compiles the Rust library for the selected target.
+Xcode passes the architectures to build in `ARCHS`.
+It's either `x86_64` for simulator builds on Intel Mac hardware or `arm64`.
+If it's `arm64` it can be either the simulator or an actual hardware target.
+Those differ, but we can know which is which from what's in `LLVM_TARGET_TRIPLE_SUFFIX` and select the right Rust target.
+
+## The code parts
+
+We now have an Xcode project that builds our Rust library and links against it.
+We now need to use this library!
+
+Swift speaks Objective-C, which is an extension to C,
+but we need to tell it about the things available.
+In C land that's done with a header.
+Let's create a new file, select the "Header File" template and name it `FfiBridge.h`.
+This will create a new file with this content:
+
+```c
+#ifndef FfiBridge_h
+#define FfiBridge_h
+
+
+#endif /* FfiBridge_h */
+```
+
+Here we need to add the definition of our function.
+As a reminder this is its definition in Rust:
+
+```rust
+extern "C" fn shipping_rust_addition(a: c_int, b: c_int) -> c_int;
+```
+
+This translates to the following in C:
+
+```c
+int shipping_rust_addition(int a, int b);
+```
+
+Add that line between the `#define` and `#endif` lines.
+Xcode doesn't know about that file yet, so once more into the `Build Settings` of the target.
+Search for `Objective-C Bridging Header` and set it to `$(PROJECT_DIR)/ShippingRust/FfiBridge.h`.
+In `Build Phases` add a new `Header Phase`.
+There you add the `FfiBridge.h` as well.
+
+If it now all compiles we're finally ready to use our Rust library.
+
+Open up `ContentView.swift` and change the code to call your Rust library:
+
+```
+struct ContentView: View {
+    var body: some View {
+        Text("Hello, world! \(shipping_rust_addition(30, 1))")
+            .padding()
+    }
+}
+```
+
+We simply interpolate the result of `shipping_rust_addition(30, 1)` into the string displayed.
+
+Once we compile and run it in the simulator we succeeded:
+
+> Show a non-interactive message to the user with data from a Rust library.
+
+![iOS simulator running our application showing "Hello, world! 31"](https://tmp.fnordig.de/blog/2022/ios-simulator-helloworld31.png)
+
+Compiling for any iOS device should work just as well.
+
+## The next steps
+
+This was a lot of setup for calling one simple function.
+Luckily this is a one-time setup. From here on you can extend your Rust library, define them in the header file and call them from Swift.
+If you go that route you should really start using [cbindgen] to generate that header file automatically for you.
+
+For Glean we're stepping away from manually writing our FFI functions.
+We're instead migrating our code base to use [UniFFI].
+UniFFI will generate the C API from an API definitions file and also comes with a bit of runtime code to handle conversion between Rust, C and Swift data types for us.
+
```

Comment:
To me this seems an abrupt end. I'd like an off-ramp, a concluding summary to ease me back out. (but again, that's just me)